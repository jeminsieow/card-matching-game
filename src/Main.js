import { StatusBar } from "expo-status-bar";
import React, { useState, useEffect, useRef } from "react";
import { Text, TouchableOpacity, StyleSheet } from "react-native";
import styled from "styled-components/native";
import { Card } from "./components";

const Container = styled.View`
  flex: 1;
  background-color: #39353c;
  padding: 20px;
`;

const Header = styled.View`
  padding-top: 50px;
  flex-direction: row;
  justify-content: space-between;
`;

const RestartButton = styled.Text`
  padding-left: 20px;
  font-size: 20px;
  color: #1b98f2;
`;

const StepsText = styled.Text`
  padding-right: 20px;
  font-size: 30px;
  color: white;
`;

const StepsNumber = styled.Text`
  font-size: 35px;
  color: #1b98f2;
`;

const CardTable = styled.View`
  padding-top: 5px;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  flex: 1;
  justify-content: space-between;
`;

const CardContainer = styled.View`
  padding: 5px;
`;

const CARD_PAIRS_VALUE = [1, 2, 3, 4, 5, 6];

// Fisher-Yates shuffle algorithm
function shuffleCards(array) {
  const length = array.length;
  for (let i = length; i > 0; i--) {
    const randomIndex = Math.floor(Math.random() * i);
    const currentIndex = i - 1;
    const temp = array[currentIndex];
    array[currentIndex] = array[randomIndex];
    array[randomIndex] = temp;
  }
  console.log(array);
  return array;
}

export default function Main() {
  const [cards, setCards] = useState(
    shuffleCards.bind(null, CARD_PAIRS_VALUE.concat(CARD_PAIRS_VALUE))
  );
  const [openCards, setOpenCards] = useState([]);
  const [clearedCards, setClearedCards] = useState({});
  const [isCardsDisabled, setIsCardsDisabled] = useState(false);
  const [steps, setSteps] = useState(0);
  const timeout = useRef(null);

  const disableCards = () => {
    setIsCardsDisabled(true);
  };
  const enableCards = () => {
    setIsCardsDisabled(false);
  };

  // Check if both cards have the same value. If they do, mark them inactive
  const evaluateCards = () => {
    const [first, second] = openCards;
    enableCards();
    if (cards[first] === cards[second]) {
      setClearedCards((prev) => ({ ...prev, [cards[first]]: true }));
      setOpenCards([]);
      return;
    }
    // Flip cards after a 500ms duration
    timeout.current = setTimeout(() => {
      setOpenCards([]);
    }, 500);
  };

  const handleCardPress = (index) => {
    setSteps(steps + 1);
    console.log(index, steps)
    if (openCards.length === 1) {
      setOpenCards((prev) => [...prev, index]);
      disableCards();
    } else {
      // If two cards are already open, we cancel timeout set for flipping cards back
      clearTimeout(timeout.current);
      setOpenCards([index]);
    }
  };

  useEffect(() => {
    if (openCards.length === 2) {
      setTimeout(evaluateCards, 500);
    }
  }, [openCards]);

  useEffect(() => {
    console.log("openCards: " + openCards);
    console.log("clearedCards: ");
    console.log(clearedCards);
  }, [openCards, clearedCards]);

  const isCardFlipped = (index) => {
    return openCards.includes(index);
  };

  const isCardCleared = (card) => {
    return clearedCards[card];
  };

  const resetGame = () => {
    setCards(
      shuffleCards.bind(null, CARD_PAIRS_VALUE.concat(CARD_PAIRS_VALUE))
    );
    setClearedCards({});
    setOpenCards([]);
    setSteps(0);
  };

  return (
    <Container>
      <Header>
        <TouchableOpacity onPress={resetGame}>
          <RestartButton>Restart</RestartButton>
        </TouchableOpacity>
        <StepsText>STEPS: <StepsNumber>{steps}</StepsNumber></StepsText>
      </Header>

      <CardTable>
        {cards.map((card, index) => {
          // return <Text key={index}>{card}</Text>;
          return (
            <CardContainer key={index}>
              <Card
                onPress={handleCardPress}
                card={card}
                index={index}
                isDisabled={isCardsDisabled}
                isCleared={isCardCleared(card)}
                isFlipped={isCardFlipped(index)}
              />
            </CardContainer>
          );
        })}
      </CardTable>

      <StatusBar style="auto" />
    </Container>
  );
}
