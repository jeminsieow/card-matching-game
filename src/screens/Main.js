import { StatusBar } from "expo-status-bar";
import React, { useState, useEffect, useRef } from "react";
import { Alert, FlatList } from "react-native";
import styled from "styled-components/native";
import { Card, Header } from "../components";
import { shuffleCards, getRandomNumbers } from "../utilities/functions";
import { CARD_PAIRS_VALUE, COLOR_DARK_GREY } from "../utilities/constants";

const Container = styled.View`
  flex: 1;
  background-color: ${COLOR_DARK_GREY};
  padding: 20px;
  align-items: center;
`;

export default function Main() {
  const [cards, setCards] = useState(CARD_PAIRS_VALUE)
  const [openCards, setOpenCards] = useState([]);
  const [clearedCards, setClearedCards] = useState([]);
  const [isCardsDisabled, setIsCardsDisabled] = useState(false);
  const [steps, setSteps] = useState(0);
  const timeout = useRef(null);
  // Prevent user from pressing more than 2 cards
  const disableCards = () => {
    setIsCardsDisabled(true);
  };
  const enableCards = () => {
    setIsCardsDisabled(false);
  };

  // Check if both cards have the same value. If they do, mark them cleared
  const evaluateCards = () => {
    const [first, second] = openCards;
    enableCards();
    if (cards[first] === cards[second]) {
      setClearedCards((prev) => [...prev, cards[first]]);
      setOpenCards([]);
      return;
    }
    // Flip cards after 1 second
    timeout.current = setTimeout(() => {
      setOpenCards([]);
    }, 1000);
  };

  const handleCardPress = (index) => {
    setSteps(steps + 1);
    if (openCards.length === 1) {
      setOpenCards((prev) => [...prev, index]);
      disableCards();
    } else {
      clearTimeout(timeout.current);
      setOpenCards([index]);
    }
  };

  const isCardFlipped = (index) => {
    return openCards.includes(index);
  };

  const isCardCleared = (card) => {
    return clearedCards.includes(card);
  };

  // Get a new array of numbers, and shuffle the array
  const resetGame = () => {
    setCards(shuffleCards(getRandomNumbers()));
    setClearedCards([]);
    setOpenCards([]);
    setSteps(0);
    setIsCardsDisabled(false);
  };

  const checkGameEnd = () => {
    if (clearedCards.length === CARD_PAIRS_VALUE.length / 2) {
      Alert.alert(
        "Congratulations!",
        `You have won this game in ${steps} steps!`,
        [{ text: "Try another round", onPress: () => resetGame() }]
      );
    }
  };

  const renderCard = ({ item, index }) => {
    return (
      <Card
        onPress={handleCardPress}
        card={item}
        index={index}
        isDisabled={isCardsDisabled}
        isCleared={isCardCleared(item)}
        isFlipped={isCardFlipped(index)}
      />
    );
  };

  useEffect(() => {
    if (openCards.length === 2) {
      setTimeout(evaluateCards, 100);
    }
  }, [openCards]);

  useEffect(() => {
    checkGameEnd();
  }, [clearedCards]);

  return (
    <Container>
      <Header onPressResetGame={resetGame} steps={steps} />
      <FlatList
        data={cards}
        numColumns={3}
        renderItem={renderCard}
        keyExtractor={(_, index) => index}
        showsVerticalScrollIndicator={false}
        scrollEnabled={false}
      />
      <StatusBar style="auto" />
    </Container>
  );
}
