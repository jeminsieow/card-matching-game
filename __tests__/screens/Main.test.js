import React from "react";
import Main from "../../src/screens/Main";
import { shallow } from "enzyme";

jest.useFakeTimers();
jest.mock("react-native/Libraries/Animated/src/NativeAnimatedHelper");
jest.mock("../../src/utilities/functions");

describe("Main component", () => {
  it("renders correctly", () => {
    const wrapper = shallow(<Main />);
    expect(wrapper).toMatchSnapshot();
  });
});
