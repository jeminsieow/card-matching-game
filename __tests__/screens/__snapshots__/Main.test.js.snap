// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component renders correctly 1`] = `
<Styled(View)>
  <Header
    onPressResetGame={[Function]}
    steps={0}
  />
  <FlatList
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={3}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
  <ExpoStatusBar
    style="auto"
  />
</Styled(View)>
`;
